aries_cloudagent/transport/inbound/queue/kafka.py

```diff
@@ -0,0 +1,264 @@
+"""Redis inbound transport."""
+import asyncio
+import os
+import msgpack
+import json
+import logging
+import pathlib
+
+from aiokafka import AIOKafkaConsumer, AIOKafkaProducer, ConsumerRebalanceListener
+from aiokafka.errors import OffsetOutOfRangeError
+from collections import Counter
+from threading import Thread
+
+from ....core.profile import Profile
+
+from ...wire_format import DIDCOMM_V0_MIME_TYPE, DIDCOMM_V1_MIME_TYPE
+
+from ..manager import InboundTransportManager
+
+from .base import BaseInboundQueue, InboundQueueConfigurationError
+
+
+class RebalanceListener(ConsumerRebalanceListener):
+    """Listener to control actions before and after rebalance."""
+
+    def __init__(self, consumer, local_state):
+        """Initialize RebalanceListener."""
+        self.consumer = consumer
+        self.local_state = local_state
+
+    async def on_partitions_revoked(self, revoked):
+        """Triggered on partitions revocation."""
+        self.local_state.dump_local_state()
+
+    async def on_partitions_assigned(self, assigned):
+        """Triggered on partitions assigned."""
+        self.local_state.load_local_state(assigned)
+        for tp in assigned:
+            last_offset = self.local_state.get_last_offset(tp)
+            if last_offset < 0:
+                await self.consumer.seek_to_beginning(tp)
+            else:
+                self.consumer.seek(tp, last_offset + 1)
+
+
+class LocalState:
```

Comment:
How do we see storage of state locally scaling? I think this would be an obstacle if we wanted to coordinate multiple ACA-Py instances pulling from the same Kafka topic.

Reply:
My intention behind `LocalState` implementation was to ensure `exactly-once` consumption of messages. As I tested this implementation with `AIOKafkaConsumer` scaling up/down (joining or dropping from the group), I found that `exactly-once` semantics would break momentarily for a few messages when rebalancing takes place. Since then, I have tried 2 other approaches, one was to use `redis` as offset storage and the other was to use Kafka's `commit` and `committed` functions and also tie these to `ConsumerRebalanceListener`. Both these work perfectly during my testing and I have adopted the second approach (not relying on `redis`) and updated my implementation accordingly.